Execution Client is offline - timeout check

Teku 22.9.1 (plus Besu 22.7.3).

I am occasionally getting `Execution Client is offline` errors, very swiftly followed by `Execution Client is online`, within a few tens of milliseconds.

Two questions:

1.  Do we need the stack trace here, or can we simplify?
2.  Is the timeout too aggressive? My execution client is not offline, and not struggling afaik. If the timeout were, say, 0.5s longer I would never see these. What are the factors used to set the timeout?

```
2022-09-25 12:56:27.397 ERROR - Execution Client is offline
java.util.concurrent.TimeoutException: null
at java.util.concurrent.CompletableFuture$Timeout.run(CompletableFuture.java:2874) ~[?:?]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
at java.lang.Thread.run(Thread.java:833) ~[?:?]
2022-09-25 12:56:27.398 ERROR - An error occurred while querying remote transition configuration
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Invalid remote response: TimeoutException
at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315) ~[?:?]
at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320) ~[?:?]
at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649) ~[?:?]
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147) ~[?:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147) ~[teku-infrastructure-async-22.9.1.jar:22.9.1]
at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863) ~[?:?]
at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841) ~[?:?]
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162) ~[?:?]
at java.util.concurrent.CompletableFuture$Timeout.run(CompletableFuture.java:2874) ~[?:?]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.IllegalArgumentException: Invalid remote response: TimeoutException
at com.google.common.base.Preconditions.checkArgument(Preconditions.java:220) ~[guava-31.1-jre.jar:?]
at tech.pegasys.teku.ethereum.executionlayer.ExecutionLayerManagerImpl.unwrapResponseOrThrow(ExecutionLayerManagerImpl.java:504) ~[teku-ethereum-executionlayer-22.9.1.jar:22.9.1]
at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646) ~[?:?]
... 14 more
2022-09-25 12:56:27.426 INFO  - Execution Client is online

...

2022-09-25 14:42:27.499 ERROR - Execution Client is offline
java.util.concurrent.TimeoutException: null
...
2022-09-25 14:42:27.636 INFO  - Execution Client is online

...

2022-09-25 19:39:27.570 ERROR - Execution Client is offline
java.util.concurrent.TimeoutException: null
...
2022-09-25 19:39:27.987 INFO  - Execution Client is online

...

2022-09-25 20:52:27.583 ERROR - Execution Client is offline
java.util.concurrent.TimeoutException: null
...
2022-09-25 20:52:28.012 INFO  - Execution Client is online
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,6 +20,7 @@ For information on changes in released versions of Teku, see the [releases page]
 ### Breaking Changes
 
 ### Additions and Improvements
-- Filter out unknown validators when sending validator registrations to the builder network
+- Improve Execution Layer error logging
 
 ### Bug Fixes
+- Filter out unknown validators when sending validator registrations to the builder network

--- a/ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/web3j/Web3JClient.java
+++ b/ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/web3j/Web3JClient.java
@@ -85,38 +85,25 @@ public abstract class Web3JClient {
             });
   }
 
-  private boolean isAuthenticationException(final Throwable exception) {
-    if (!(exception instanceof ClientConnectionException)) {
-      return false;
-    }
-    final String message = exception.getMessage();
-    return message.contains("received: 401") || message.contains("received: 403");
-  }
-
-  private boolean isTimeoutException(final Throwable exception) {
-    return exception instanceof TimeoutException;
-  }
-
   protected void handleError(final Throwable error) {
     handleError(error, false);
   }
 
   protected void handleError(final Throwable error, final boolean couldBeAuthError) {
     final long timeNow = timeProvider.getTimeInMillis().longValue();
-    lastError.accumulateAndGet(
-        timeNow,
-        (lastErrorTime, givenErrorTimeUpdate) -> {
-          if (lastErrorTime == NO_ERROR_TIME
-              || givenErrorTimeUpdate - lastErrorTime > ERROR_REPEAT_DELAY_MILLIS) {
-            if (isTimeoutException(error)) {
-              eventLog.executionClientRequestTimedOut();
-            } else {
-              eventLog.executionClientRequestFailed(error, couldBeAuthError);
-            }
-            return givenErrorTimeUpdate;
-          }
-          return lastErrorTime;
-        });
+    final long maybeUpdatedTime =
+        lastError.accumulateAndGet(
+            timeNow,
+            (lastErrorTime, givenErrorTimeUpdate) -> {
+              if (lastErrorTime == NO_ERROR_TIME
+                  || givenErrorTimeUpdate - lastErrorTime > ERROR_REPEAT_DELAY_MILLIS) {
+                return givenErrorTimeUpdate;
+              }
+              return lastErrorTime;
+            });
+    if (maybeUpdatedTime == timeNow) {
+      logExecutionClientError(error, couldBeAuthError);
+    }
   }
 
   protected void handleSuccess() {
@@ -140,5 +127,21 @@ public abstract class Web3JClient {
 
   public boolean isWebsocketsClient() {
     return false;
+  }
+
+  private boolean isAuthenticationException(final Throwable exception) {
+    if (!(exception instanceof ClientConnectionException)) {
+      return false;
+    }
+    final String message = exception.getMessage();
+    return message.contains("received: 401") || message.contains("received: 403");
+  }
+
+  private void logExecutionClientError(final Throwable error, final boolean couldBeAuthError) {
+    if (error instanceof TimeoutException) {
+      eventLog.executionClientRequestTimedOut();
+    } else {
+      eventLog.executionClientRequestFailed(error, couldBeAuthError);
+    }
   }
 }

--- a/ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/web3j/Web3JClient.java
+++ b/ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/web3j/Web3JClient.java
@@ -17,6 +17,7 @@ import static tech.pegasys.teku.infrastructure.exceptions.ExceptionUtil.getMessa
 
 import java.io.IOException;
 import java.time.Duration;
+import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicLong;
 import org.web3j.protocol.Web3j;
 import org.web3j.protocol.Web3jService;
@@ -28,16 +29,19 @@ import tech.pegasys.teku.infrastructure.logging.EventLogger;
 import tech.pegasys.teku.infrastructure.time.TimeProvider;
 
 public abstract class Web3JClient {
+
   private static final int ERROR_REPEAT_DELAY_MILLIS = 30 * 1000;
   private static final int NO_ERROR_TIME = -1;
+  private static final long STARTUP_LAST_ERROR_TIME = 0;
+
   private final EventLogger eventLog;
   private final TimeProvider timeProvider;
   private Web3jService web3jService;
   private Web3j eth1Web3j;
 
-  // Default to the provider being offline at startup so we log when it is first available
-  // but uses a very old value to make sure we log if the first request fails
-  private final AtomicLong lastError = new AtomicLong(0);
+  // Default to the provider having a previous failure at startup so we log when it is first
+  // available but uses a very old value to make sure we log if the first request fails
+  private final AtomicLong lastError = new AtomicLong(STARTUP_LAST_ERROR_TIME);
   private boolean initialized = false;
 
   protected Web3JClient(final EventLogger eventLog, final TimeProvider timeProvider) {
@@ -66,7 +70,8 @@ public abstract class Web3JClient {
         .handle(
             (response, exception) -> {
               if (exception != null) {
-                handleError(exception, isAuthenticationException(exception));
+                final boolean couldBeAuthError = isAuthenticationException(exception);
+                handleError(exception, couldBeAuthError);
                 return Response.withErrorMessage(getMessageOrSimpleName(exception));
               } else if (response.hasError()) {
                 final String errorMessage =
@@ -88,23 +93,33 @@ public abstract class Web3JClient {
     return message.contains("received: 401") || message.contains("received: 403");
   }
 
+  private boolean isTimeoutException(final Throwable exception) {
+    return exception instanceof TimeoutException;
+  }
+
   protected void handleError(final Throwable error) {
     handleError(error, false);
   }
 
   protected void handleError(final Throwable error, final boolean couldBeAuthError) {
-    final long errorTime = lastError.get();
-    if (errorTime == NO_ERROR_TIME
-        || timeProvider.getTimeInMillis().longValue() - errorTime > ERROR_REPEAT_DELAY_MILLIS) {
-      if (lastError.compareAndSet(errorTime, timeProvider.getTimeInMillis().longValue())) {
-        eventLog.executionClientIsOffline(error, couldBeAuthError);
+    final long lastErrorTime = lastError.get();
+    final long timeNow = timeProvider.getTimeInMillis().longValue();
+    if (lastErrorTime == NO_ERROR_TIME || timeNow - lastErrorTime > ERROR_REPEAT_DELAY_MILLIS) {
+      lastError.set(timeNow);
+      if (isTimeoutException(error)) {
+        eventLog.executionClientRequestTimedOut();
+      } else {
+        eventLog.executionClientRequestFailed(error, couldBeAuthError);
       }
     }
   }
 
   protected void handleSuccess() {
-    if (lastError.getAndUpdate(x -> NO_ERROR_TIME) != NO_ERROR_TIME) {
+    final long lastErrorTime = lastError.getAndUpdate(x -> NO_ERROR_TIME);
+    if (lastErrorTime == STARTUP_LAST_ERROR_TIME) {
       eventLog.executionClientIsOnline();
+    } else if (lastErrorTime != NO_ERROR_TIME) {
+      eventLog.executionClientRecovered();
     }
   }
 

--- a/ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/web3j/Web3JClient.java
+++ b/ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/web3j/Web3JClient.java
@@ -102,16 +102,21 @@ public abstract class Web3JClient {
   }
 
   protected void handleError(final Throwable error, final boolean couldBeAuthError) {
-    final long lastErrorTime = lastError.get();
     final long timeNow = timeProvider.getTimeInMillis().longValue();
-    if (lastErrorTime == NO_ERROR_TIME || timeNow - lastErrorTime > ERROR_REPEAT_DELAY_MILLIS) {
-      lastError.set(timeNow);
-      if (isTimeoutException(error)) {
-        eventLog.executionClientRequestTimedOut();
-      } else {
-        eventLog.executionClientRequestFailed(error, couldBeAuthError);
-      }
-    }
+    lastError.accumulateAndGet(
+        timeNow,
+        (lastErrorTime, givenErrorTimeUpdate) -> {
+          if (lastErrorTime == NO_ERROR_TIME
+              || givenErrorTimeUpdate - lastErrorTime > ERROR_REPEAT_DELAY_MILLIS) {
+            if (isTimeoutException(error)) {
+              eventLog.executionClientRequestTimedOut();
+            } else {
+              eventLog.executionClientRequestFailed(error, couldBeAuthError);
+            }
+            return givenErrorTimeUpdate;
+          }
+          return lastErrorTime;
+        });
   }
 
   protected void handleSuccess() {

--- a/ethereum/executionclient/src/test/java/tech/pegasys/teku/ethereum/executionclient/web3j/Web3JClientTest.java
+++ b/ethereum/executionclient/src/test/java/tech/pegasys/teku/ethereum/executionclient/web3j/Web3JClientTest.java
@@ -45,6 +45,8 @@ public class Web3JClientTest {
   private static final URI ENDPOINT = URI.create("");
   private static final Duration DEFAULT_TIMEOUT = Duration.ofMinutes(1);
 
+  private final EventLogger eventLog = mock(EventLogger.class);
+
   @SuppressWarnings("unused")
   static Stream<Arguments> getClientInstances() {
     final TimeProvider timeProvider = StubTimeProvider.withTimeInSeconds(1000);
@@ -90,7 +92,7 @@ public class Web3JClientTest {
   @ParameterizedTest
   @MethodSource("getClientInstances")
   void shouldTimeoutIfResponseNotReceived(final ClientFactory clientFactory) throws Exception {
-    final Web3JClient client = clientFactory.create();
+    final Web3JClient client = clientFactory.create(eventLog);
     Request<Void, VoidResponse> request =
         new Request<>("test", new ArrayList<>(), client.getWeb3jService(), VoidResponse.class);
     when(client.getWeb3jService().sendAsync(request, VoidResponse.class))
@@ -102,12 +104,12 @@ public class Web3JClientTest {
     SafeFutureAssert.assertThatSafeFuture(result).isCompleted();
     final Response<Void> response = SafeFutureAssert.safeJoin(result);
     assertThat(response.getErrorMessage()).isEqualTo(TimeoutException.class.getSimpleName());
+    verify(eventLog).executionClientRequestTimedOut();
   }
 
   @ParameterizedTest
   @MethodSource("getClientInstances")
   void shouldLogOnFirstSuccess(final ClientFactory clientFactory) throws Exception {
-    final EventLogger eventLog = mock(EventLogger.class);
     final Web3JClient client = clientFactory.create(eventLog);
     Request<Void, VoidResponse> request =
         new Request<>("test", new ArrayList<>(), client.getWeb3jService(), VoidResponse.class);
@@ -123,29 +125,23 @@ public class Web3JClientTest {
   @ParameterizedTest
   @MethodSource("getClientInstances")
   void shouldReportFailureAndRecovery(final ClientFactory clientFactory) throws Exception {
-    final EventLogger eventLog = mock(EventLogger.class);
     final Web3JClient client = clientFactory.create(eventLog);
     Request<Void, VoidResponse> request =
         new Request<>("test", new ArrayList<>(), client.getWeb3jService(), VoidResponse.class);
-    final Throwable error = new TimeoutException();
+    final Throwable error = new IllegalStateException("oopsy");
     when(client.getWeb3jService().sendAsync(request, VoidResponse.class))
         .thenReturn(SafeFuture.failedFuture(error))
         .thenReturn(SafeFuture.completedFuture(new VoidResponse()));
 
     Waiter.waitFor(client.doRequest(request, DEFAULT_TIMEOUT));
 
-    verify(eventLog).executionClientIsOffline(error, false);
+    verify(eventLog).executionClientRequestFailed(error, false);
 
     Waiter.waitFor(client.doRequest(request, DEFAULT_TIMEOUT));
-    verify(eventLog).executionClientIsOnline();
+    verify(eventLog).executionClientRecovered();
   }
 
   private interface ClientFactory {
-
-    default Web3JClient create() {
-      return create(mock(EventLogger.class));
-    }
-
     Web3JClient create(EventLogger eventLog);
   }
 }

--- a/infrastructure/logging/src/main/java/tech/pegasys/teku/infrastructure/logging/EventLogger.java
+++ b/infrastructure/logging/src/main/java/tech/pegasys/teku/infrastructure/logging/EventLogger.java
@@ -34,6 +34,9 @@ public class EventLogger {
   public static final EventLogger EVENT_LOG =
       new EventLogger(LoggingConfigurator.EVENT_LOGGER_NAME);
 
+  private static final String EXECUTION_CLIENT_READINESS_USER_REMINDER =
+      "Make sure the Execution Client is online and ready.";
+
   @SuppressWarnings("PrivateStaticFinalLoggers")
   private final Logger log;
 
@@ -129,18 +132,28 @@ public class EventLogger {
     info("Beacon chain syncing complete, waiting for Execution Client", Color.YELLOW);
   }
 
-  public void executionClientIsOffline(final Throwable error, final boolean couldBeAuthError) {
+  public void executionClientIsOnline() {
+    info("Execution Client is online", Color.GREEN);
+  }
+
+  public void executionClientRequestFailed(final Throwable error, final boolean couldBeAuthError) {
     error(
-        "Execution Client is offline"
+        "Execution Client request failed. "
             + (couldBeAuthError
-                ? ". Check the same JWT secret is configured for Teku and the execution client."
-                : ""),
+                ? "Check the same JWT secret is configured for Teku and the Execution Client."
+                : EXECUTION_CLIENT_READINESS_USER_REMINDER),
         Color.RED,
         error);
   }
 
-  public void executionClientIsOnline() {
-    info("Execution Client is online", Color.GREEN);
+  public void executionClientRequestTimedOut() {
+    error(
+        "Execution Client request timed out. " + EXECUTION_CLIENT_READINESS_USER_REMINDER,
+        Color.RED);
+  }
+
+  public void executionClientRecovered() {
+    info("Execution Client request succeeded after a previous failure", Color.GREEN);
   }
 
   public void builderIsOffline(final String errorMessage) {

